### PR TITLE
Filtering UI for passwords

### PR DESCRIPTION
**Pull request by beaucollins**
Allows filtering of tokens by searching the token name and token issuer. Does a case-insensitive test of wether either name or issuer contain the user provided string.

When a filter is applied sorting is disabled during _Edit_ mode.

See #113 

Screencast: https://cldup.com/A8wKuiPHiD.mov

screenshot

### CODE REVIEW

File: Authenticator/Source/TokenRowModel.swift [hunk 1]
```diff
@@ -31,14 +31,15 @@ struct TokenRowModel: Identifiable {
...
+    let canSort: Bool
     let buttonAction: Action
     let selectAction: Action
     let editAction: Action
     let deleteAction: Action
 
     private let identifier: NSData
 
-    init(persistentToken: PersistentToken, displayTime: DisplayTime) {
+    init(persistentToken: PersistentToken, displayTime: DisplayTime, canSort sortable: Bool = true) {
```
Comment (beaucollins) [inline Authenticator/Source/TokenRowModel.swift]: Fixed

**Comment by beaucollins** [comment]
@mattrubin I would love a review of this!

File: Authenticator/Source/TokenListViewController.swift [hunk 2]
```diff
@@ -232,3 +257,117 @@ extension TokenListViewController {
...
+        addSubview(textField)
+    }
+
+    override func intrinsicContentSize() -> CGSize {
+        return ring.frame.union(textField.frame).size
+    }
+
+    override func alignmentRectForFrame(frame: CGRect) -> CGRect {
+        return super.alignmentRectForFrame(frame)
+    }
```
Comment (mattrubin) [inline Authenticator/Source/TokenListViewController.swift]: What purpose does this function override serve?

File: Authenticator/Source/TableDiff.swift [hunk 3]
```diff
@@ -80,7 +80,7 @@ private func changes<T>(from oldItems: [T], to newItems: [T], hasSameIdentity: (
             var y = x - k
             while x < oldItems.count && y < newItems.count
                 && hasSameIdentity(oldItems[x], newItems[y]) {
-                    if !isEqual(oldItems[x], newItems[y]) {
+                    if !isEqual(oldItems[x], newItems[y]) && (x != y) {
```
Comment (beaucollins) [inline Authenticator/Source/TableDiff.swift]: II was experiencing crashes on my branch when sorting the last item upwards. Remove this additional logic check and you'll see it. I can post the error trace but it had to do with a single row having more than one operation.

I believe the underlying problem may have to do with the map/filter of the filtered tokens but never got to the bottom of it.

File: Authenticator/Source/TokenListViewController.swift [hunk 4]
```diff
@@ -25,7 +25,7 @@
 
 import UIKit
 
-class TokenListViewController: UITableViewController {
+class TokenListViewController: UITableViewController, UITextFieldDelegate {
```
Comment (beaucollins) [inline Authenticator/Source/TokenListViewController.swift]: I believe I added it due to the compiler telling me to. I can check.

File: Authenticator/Source/TokenListViewController.swift [hunk 5]
```diff
@@ -103,6 +112,13 @@ class TokenListViewController: UITableViewController {
     override func viewWillAppear(animated: Bool) {
         super.viewWillAppear(animated)
 
+        NSNotificationCenter
+            .defaultCenter()
+            .addObserver(self,
+                         selector: #selector(onTextFieldUpdated),
+                         name: UITextFieldTextDidChangeNotification,
+                         object: searchBar.textField)
```
Comment (beaucollins) [inline Authenticator/Source/TokenListViewController.swift]: Will do.

File: Authenticator/Source/TokenListViewController.swift [hunk 6]
```diff
@@ -232,3 +257,117 @@ extension TokenListViewController {
         }
     }
 }
+
+extension TokenListViewController: UITextViewDelegate {
```
Comment (beaucollins) [inline Authenticator/Source/TokenListViewController.swift]: It looks like this is necessary anyway.

File: Authenticator/Source/TokenListViewController.swift [hunk 7]
```diff
@@ -182,17 +197,24 @@ extension TokenListViewController {
...
-    private func updateTableViewWithChanges(changes: [Change]) {
+    private func updateTableViewWithChanges(changes: [Change],
+                                            whileFiltering isFiltering: Bool = false) {
         // TODO: Scroll to a newly added token (added at the bottom)
         if changes.isEmpty || preventTableViewAnimations {
             return
         }
 
+        if isFiltering {
+            tableView.reloadData()
```
Comment (beaucollins) [inline Authenticator/Source/TokenListViewController.swift]: Yeah, it was unnecessary. When searching I wanted to see my results _right away_.

**Comment by beaucollins** [comment]
I'm a pretty big fan of the Elm architecture and I like how it's set up here. In my "day job" I've been doing a lot of react/redux. I've become somewhat of a functional programming junky lately so I find the pattern very pleasant to develop with.

One drawback I discovered in this app is the timer forcing a reevaluation of the state model every second or so and since there's no memoization or caching of the view-model -> ui states there's a lot of view redrawing that may not be necessary in the UITableView.

Here's a specific example that can be reproduced in `master`:

The OTP ring progress view updates using the CADisplayLink which by default runs on the main loop. You'll notice then that while scrolling the UITableView the ring no longer updates its progress due to the scrolling animations happening on a different runLoop.

At the end of the day, the ring not animating during scrolling is not a big deal but it did bug me a bit. I modified the CADisplayLink line to run `tick` on both the main loop and the `UITrackingRunLoopMode` (by using `NSRunLoopCommonModes`). The ring now animated even while scrolling but the `UIScrollView` went bonkers. My assumption is that the `UITableView` is getting updated during scrolling due to the `tick` causing the whole view model to reevaluate. If I comment out `updateTableViewWithChanges(changes)` scrolling goes back to normal *and* the ring continues to update its progress.

Something like Redux's [reselect](https://github.com/reactjs/reselect) library could be useful in this case.

**Comment by beaucollins** [comment]
> (...and apologies for the delayed review.

No apologies necessary! Thank you for your time and for making Authenticator. It was the perfect replacement for Google's. I've been a happy user for quite some time now.

**Comment by beaucollins** [comment]
Some insights regarding the `TableDiff` change I made (an have since reverted):

If I have more than a screen's worth of tokens this crash happens when reordering the last item in the table view:

```
2016-10-10 09:40:52.195995 Authenticator[3852:1015298] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Attempt to create two animations for cell'
```

From my tests now it happens on this branch as well as `master` so I'll file as a separate issue.

File: Authenticator/Source/TokenList.swift [hunk 8]
```diff
@@ -91,6 +111,9 @@ extension TokenList {
         case TokenChangeSucceeded([PersistentToken])
         case UpdateTokenFailed(ErrorType)
         case DeleteTokenFailed(ErrorType)
+
+        case Filter(String)
+        case ClearFilter()
```
Comment (beaucollins) [inline Authenticator/Source/TokenList.swift]: Fixed

File: Authenticator/Source/TokenRowModel.swift [hunk 9]
```diff
@@ -40,7 +40,7 @@ struct TokenRowModel: Identifiable {
     private let identifier: NSData
 
     init(persistentToken: PersistentToken, displayTime: DisplayTime,
-         canSort sortable: Bool = true) {
+         conReorder reorderable: Bool = true) {
```
Comment (beaucollins) [inline Authenticator/Source/TokenRowModel.swift]: 😞

File: Authenticator/Source/TokenListViewModel.swift [hunk 10]
```diff
@@ -28,4 +28,11 @@ import Foundation
 struct TokenListViewModel {
     let rowModels: [TokenRowModel]
     let ringProgress: Double?
+    let totalTokens: Int
+    var isFiltering: Bool {
+        return totalTokens > rowModels.count
```
Comment (mattrubin) [inline Authenticator/Source/TokenListViewModel.swift]: Could this have unexpected side effects if the filter applies to (and shows) every available token? (That is, if the filter is active, but all tokens pass the filter so `rowModels.count` still equals `totalTokens`.)

File: Authenticator/Source/SearchField.swift [hunk 11]
```diff
@@ -101,6 +101,25 @@ class SearchField: UIView {
...
 
+// MARK: TokenListPresenter
+extension SearchField {
+    func updateWithViewModel(viewModel: TokenList.ViewModel) {
+        if let ringProgress = viewModel.ringProgress {
+            ring.progress = ringProgress
+        }
+        // Show the countdown ring only if a time-based token is active
+        textField.leftViewMode = viewModel.ringProgress != nil ? .Always : .Never
+        textField.borderStyle = viewModel.hasTokens ? .RoundedRect : .None
```
Comment (beaucollins) [inline Authenticator/Source/SearchField.swift]: The filter UI is disabled using the same view model property on line 115.

File: Authenticator/Source/TokenListViewModel.swift (hunk 10 above)
Comment (beaucollins) [inline Authenticator/Source/TokenListViewModel.swift]: At this point the only place `isFiltering` is being used is to determine if the `UITableView` in the `TokenListViewController` should animate row changes. But you are correct that a filter `String` may be set on the `TokenList` but the `TokenListViewModel.isFiltering` will return `false` when the `TokenList.filter` matches all `PersistentToken` instances.

`TokenList` also has an `isFiltering` property that is determined by the value of the `TokenList.filter` property. I can pass that to the `TokenListViewModel` struct on initialization.

**Pull request merged**